dev/financial#82 Add getters & setters for contributionID on CRM_Core_Payment.

eileenmcnaughton opened the pull request:
Overview
----------------------------------------
Add getters & setters for various parameters CRM_Core_Payment. as part of defining & standardising the interface.

This will enable us to start to lock down what the parameters should actually be & move to a more structured form.

There should  be no immediate change  as a result of this but we would then start to encourage  processors to switch to using getters. 

Calling code could also start using setters  and in some cases (e.g. description) that would override what is currently being used, but generally would need to do BOTH for transitional reasons.

Before
----------------------------------------
'arbitrary' $params passed in with no clarify

After
----------------------------------------
The start of a move to defined properties

Technical Details
----------------------------------------
We currently have 
```
$paymentProcessor->getID();
```
- We have these -  but  this makes $params an optional field for them.
```
$paymentProcessor->getAmount();
$paymentProcessor->getCurrency();
```

& this adds
```
$paymentProcessor->getContributionID();
$paymentProcessor->getContactID();
$paymentProcessor->getContributionRecurID();
$paymentProcessor->getPaymentToken();
$paymentProcessor->getDescription(); (note this  is  'calculated' in getPaymentDescription if not set)
$paymentProcessor->getInvoiceID();
$paymentProcessor->isRecur();
$paymentProcessor->getRecurFrequencyInterval();
$paymentProcessor->getRecurFrequencyUnit();
```
Processors should also set some things
```
$paymentProcessor->setTransactionID();
$paymentProcessor->setFeeAmount();
```

There are more we could add - probably as a follow up

Comments
----------------------------------------
https://lab.civicrm.org/dev/financial/issues/82

eileenmcnaughton commented:
test this please

mattwire commented:
test this please

eileenmcnaughton commented:
test this please

seamuslee001 commented:
Finally @eileenmcnaughton :-)

mattwire commented:
@eileenmcnaughton I can see the sense in this but I feel like it's adding quite a lot of complexity without necessarily solving the original problem - ie. that we need to set contribution_id and contributionID on `$params`.  I've taken a simpler approach in my mjwshared extension and just put comments / abstraction into the getter. The payment processor still has to call the getter for it to be effective but it's simple to understand and manage - eg. `getContributionID()`: https://lab.civicrm.org/extensions/mjwshared/blob/master/CRM/Core/Payment/MJWTrait.php#L128-140

`getContactID()` is somewhat more complex: https://lab.civicrm.org/extensions/mjwshared/blob/master/CRM/Core/Payment/MJWTrait.php#L100-119

Also, historically the stripe extension ended up in quite a mess because it had loads of class properties and I think we'd risk ending up with the same in CRM_Core_Payment if we're not careful.
We already have a precedent for the "simpler way" for getCurrency(), getAmount()

eileenmcnaughton commented:
@mattwire I think long term is would be better to use getters & setters rather than passing in '$params'  because of   the arbitraryness of  a vague/undefined array. 

moving towards getters & setters does require grandfathering in though because it takes time for processors to migrate but at the point when we are adding new functions (& new api) to process payments it would be good to start to get away from a set of not-always-consistent historical parameters

eileenmcnaughton commented:
@mattwire reading this function seems like a good argument for getters & settings

```
  protected function getContactId($params) {
    // $params['contact_id'] is preferred.
    // contactID is set by: membership payment workflow
    // cms_contactID is set by: membership payment workflow when "on behalf of" / related contact is used.
    $contactId = CRM_Utils_Array::value('contactID', $params,
      CRM_Utils_Array::value('contact_id', $params,
        CRM_Utils_Array::value('cms_contactID', $params,
          CRM_Utils_Array::value('cid', $params, NULL
          ))));
    if (!empty($contactId)) {
      return $contactId;
    }

```

eileenmcnaughton commented:
Regarding the original issue - your PR just needs a test - it's not held up on this

mattwire commented:
@eileenmcnaughton Are you able to expand the scope of this PR to include the `getContactID()` function I referenced?  I think that may make it easier to accept this as the "right way to do things" as we're likely to start using it if there's more than one, and contactID tends to be pretty important!  Perhaps target 5.21 for this as we've got a lot going into 5.20

eileenmcnaughton commented:
@mattwire I've added quite a bit more to it! I do think getContactID could be cleverer  after  seeing  your one  but this PR has now become 'let's add the structure' so adding the cleverness of yours would probably make sense once this is merged.

eileenmcnaughton commented:
Also see the gitlab issue I raised for detail

mattwire commented on CRM/Core/Payment.php:
```diff
@@ -149,6 +149,338 @@ abstract class CRM_Core_Payment {
+   * @return string
+   */
+  public function getPaymentToken(): string {
+    return $this->paymentToken ?? $this->inputParams['token'];
```
Can we add a comment here that `For multi-step forms (eg. contribution/event registration with confirmation) you can set the pre_approval_parameters array on the form and it will be merged into inputParams (pre_approval_parameters['token'] would set it here).`

eileenmcnaughton replied:
I've added some more comments but not re pre_approval_parameters because I need a bigger dig to get all the comments right there & I don't want to do that in this round

mattwire commented on CRM/Core/Payment.php:
```diff
@@ -149,6 +149,338 @@ abstract class CRM_Core_Payment {
+   */
+  public function setTransactionID(string $transactionID) {
+    $this->transactionID = $transactionID;
+  }
```
`Note that this is not the gateway generated Order/Invoice reference (which can be set in civicrm_financial_trxn but references all tranactions related to a single Order/Invoice).`

eileenmcnaughton replied:
More comments added

mattwire commented on CRM/Core/Payment.php:
```diff
@@ -149,6 +149,338 @@ abstract class CRM_Core_Payment {
+   *
+   * @param string $resultCode
+   */
+  public function setResultCode(string $resultCode) {
```
See comment above - I'd prefer `get/setTrxnResultCode`

eileenmcnaughton replied:
OK - you win this one :-)

mattwire commented on CRM/Core/Payment.php:
```diff
@@ -149,6 +149,338 @@ abstract class CRM_Core_Payment {
+   */
+  public function setInvoiceID(string $invoiceID) {
+    $this->invoiceID = $invoiceID;
+  }
```
Yes. Nice and clear. Maybe later we'd move code that generates the invoice ID to `getInvoiceID` so we don't have to sprinkle it through various API / BAO functions.

eileenmcnaughton replied:
I think it should be in Order.create - I think https://lab.civicrm.org/dev/financial/issues/78

mattwire replied:
I've commented on that issue.

mattwire commented on CRM/Core/Payment.php:
```diff
@@ -1629,7 +1965,10 @@ public function subscriptionURL($entityID = NULL, $entity = NULL, $action = 'can
+  protected function getPaymentDescription($params = [], $length = 24) {
+    if ($this->getDescription()) {
+      return $this->getDescription();
+    }
```
Not sure what is best here - it seems a shame to override the useful getPaymentDescription function completely instead of just using getDescription instead of `$params['description']`.

eileenmcnaughton replied:
OK - I've altered this to move the string logic to getDescription

mattwire commented on CRM/Core/Payment/PayPalImpl.php:
```diff
@@ -903,6 +906,7 @@ public function doTransferCheckout(&$params, $component = 'contribute') {
 
     $cancelUrlString = "$cancel=1&cancel=1&qfKey={$params['qfKey']}";
     if (!empty($params['is_recur'])) {
+      // @todo use $this->getContributionID();
```
Ok. Though possibly better just to add a class level todo as we can switch over various params to getters?

eileenmcnaughton commented on CRM/Core/Payment.php:
```diff
@@ -149,6 +149,338 @@ abstract class CRM_Core_Payment {
+   * @return int
+   */
+  public function getContributionRecurID(): int {
+    return $this->contributionRecurID ?? $this->inputParams['contributionRecurID'];
```
OK - I think I know what you mean - added. NB if the contributionRecurID (or others) have varied formats I think we should handle them in here - with varying amounts of noise over time - I know your extension does that for contact ID & we should bring that in in the next round

eileenmcnaughton commented on CRM/Core/Payment.php:
```diff
@@ -149,6 +149,338 @@ abstract class CRM_Core_Payment {
+   * @var string
+   *   month|day|year
+   */
+  protected $recurFrequencyUnit;
```
I've added more - if it's not what you mean maybe do a follow up?

eileenmcnaughton commented on CRM/Core/Payment.php:
```diff
@@ -149,6 +149,338 @@ abstract class CRM_Core_Payment {
+   *
+   * @var string
+   */
+  protected $resultCode;
```
OK - I prefer resultCode though - just saying :-)

eileenmcnaughton commented on CRM/Core/Payment.php:
```diff
@@ -149,6 +149,338 @@ abstract class CRM_Core_Payment {
+   * The key thing is it is passed to the processor in lieu of card details to
+   * initiate a payment.
+   *
+   * If the token is not totally transient it would be saved to civicrm_payemnt_token.token.
```
@mattwire  I've added more docs but we should follow up on the pre_approval_parameters - which have been around since before 4.6 I think!

eileenmcnaughton commented:
@mattwire I think I've addressed your comments. Hopefully we can merge this & maybe do some follow up. Maybe we could even dig into event a little & start using some setters - although I think it has to be AND at this stage not or :-( But in time we can start to add deprecation notices if the getters have to fall back

mattwire commented on CRM/Core/Payment.php:
```diff
@@ -149,6 +149,381 @@ abstract class CRM_Core_Payment {
+  /**
+   * Additional information returned by the payment processor regarding the payment outcome.
+   *
+   * This would normally be saved in civicrm_financial_trxn.result_code.
```
Please update comment to `civicrm_financial_trxn.trxn_result_code`

mattwire commented on CRM/Core/Payment.php:
```diff
@@ -149,6 +149,381 @@ abstract class CRM_Core_Payment {
+  /**
+   * Set additional result information from gateway.
+   *
+   * This is saved to civicrm_financial_trxn.result_code.
```
`civicrm_financial_trxn.trxn_result_code`

seamuslee001 commented:
Added Merge on pass as @mattwire 's issues have been addressed
